### PR TITLE
backupccl: validate descriptors during restore

### LINKED ISSUE #84757: importccl,backupccl: use `crdb_internal.invalid_objects` to check for invalid descriptors
SQL Schema have a virtual table to check for invalid descriptors in a cluster. Restore and import are in the business of creating and ingesting descriptors, and so we should check the virtual table post restore/import for any invalid descriptor states they might have generated. This should be added to as many restore/import tests as possible, but also potentially as a sanity check during a restore/import once the descriptors have been published to an online state.

Jira issue: CRDB-17847

### PR DESCRIPTION
**Pull request by msbutler**
This patch checks that the restored descriptors are valid in two
places during the job. First, as soon as the descriptors are written in an
offline state before the restoration flow begins. This primarily ensures that
the descriptors in the backup were in a valid state, allowing the restore to
fail fast. Second, right after the descriptors return online and before the job
begins, which ensures the restore created and modified the descriptors
correctly.

Fixes https://github.com/cockroachdb/cockroach/issues/84757
Release note: None

### CODE REVIEW

File: pkg/ccl/backupccl/restore_job.go [hunk 1]
```diff
@@ -692,6 +698,31 @@ func backedUpDescriptorWithInProgressImportInto(
 	return true, nil
 }
 
+// validateRestoredDescriptors returns an err if one of the restored descriptors is invalid
+func validateRestoredDescriptors(ctx context.Context, execConfig *sql.ExecutorConfig,
```
Comment (msbutler) [inline pkg/ccl/backupccl/restore_job.go:702]: pass in ability to validate jobs?

**Comment by msbutler** [comment]
heh, it seems that check for valid descriptors right after writing them offline [causes a lot of failures](https://teamcity.cockroachdb.com/viewLog.html?buildId=7324007&problemId=4141). testing what happens if we remove the first check.

**Comment by msbutler** [comment]
all the above failures relate to the fact that full cluster restore writes system tables into a temp system table, which then should get dropped before validation. The cascade drop, however, may not get applied immediately due to https://github.com/cockroachdb/cockroach/issues/87906. This means the LoadAllDescs() call will gather temp system database table descriptors that are about to get dropped, even after the temp system database has been dropped. This seems weird and will check with schema if this behavior is expected.

File: pkg/ccl/backupccl/restore_job.go [hunk 2]
```diff
@@ -692,6 +694,32 @@ func backedUpDescriptorWithInProgressImportInto(
...
+	allDescs, err := backupresolver.LoadAllDescs(ctx, execConfig, hlc.Timestamp{WallTime: timeutil.Now().
+		UnixNano()})
+	if err != nil {
+		return err
+	}
+	targetDescs := make([]catalog.Descriptor, 0)
+	for _, desc := range allDescs {
+		if desc.Dropped() {
+			// For some reason, ValidateDescriptors fails on the temp system table
+			// when they're in the dropped stage.
```
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_job.go:708]: Do you have an example of the validation failure?

File: pkg/ccl/backupccl/restore_job.go [hunk 3]
```diff
@@ -692,6 +694,32 @@ func backedUpDescriptorWithInProgressImportInto(
...
+			continue
+		}
+	}
+	ok, invalidMsg, err := backuputils.ValidateDescriptors(ctx, targetDescs,
+		execConfig.Settings.Version.BinaryVersion())
+	if err != nil {
+		return err
+	}
+	if !ok {
+		return errors.Newf("Failed descriptor validation check: %s", invalidMsg)
```
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_job.go:718]: [nit] Here and elsewhere, I think our preferred error style is to start with a lowercase letter.  I might go with the slightly terser: "descriptor validation failed: %s"

File: pkg/ccl/backupccl/restore_planning.go [hunk 4]
```diff
@@ -1699,6 +1699,9 @@ func doRestorePlan(
 		return err
 	}
 
+	if err := validateRestoredDescriptors(ctx, p.ExecCfg()); err != nil {
+		return errors.Wrap(err, "the restoring cluster contains invalid descriptors, %s")
+	}
```
Comment (stevendanna) [inline pkg/ccl/backupccl/restore_planning.go:1704]: 🤔 I wonder if we want an option to ignore these checks.  Should an unrelated descriptor problem stop your attempt to restore?

**Review by stevendanna — CHANGES_REQUESTED** [review]
Nice to see this moving along!

File: pkg/ccl/backupccl/restore_planning.go (hunk 4 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go:1704]: i was going to ask. I think that would be a good idea.

File: pkg/ccl/backupccl/restore_job.go (hunk 2 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:708]: If I remove this skip line, all cluster restores fail with the following error msg on each temp system table  "no matching name info in draining names of dropped relation"

File: pkg/ccl/backupccl/backuputils/utils.go [hunk 5]
```diff
@@ -85,3 +93,53 @@ func AppendPaths(uris []string, tailDir ...string) ([]string, error) {
 	}
 	return retval, nil
 }
+
+// ValidateDescriptors ensures that the passed in descriptors are in a valid
+// state. The first returned parameter is true, if all descriptors are valid;
+// The second provides a description of any descriptors in an invalid state; the
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/backuputils/utils.go]: Done.

File: pkg/ccl/backupccl/backuputils/utils.go [hunk 6]
```diff
@@ -85,3 +93,53 @@ func AppendPaths(uris []string, tailDir ...string) ([]string, error) {
...
+// third returns an error if the descriptors could not be properly validated.
+// Note that this function does not check any references to the jobs table in
+// the descriptors.
+func ValidateDescriptors(
+	ctx context.Context, descriptors catalog.Descriptors, version roachpb.Version,
+) (bool, string, error) {
+	var descTable doctor.DescriptorTable
+	var namespaceTable doctor.NamespaceTable
```
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/backuputils/utils.go]: Done.

File: pkg/ccl/backupccl/restore_job.go (hunk 2 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_job.go:714]: @fqazi is it surprising that this validation check would fail on dropping descriptors?

**Comment by msbutler** [comment]
@fqazi a few questions for you:
- could confirm that i'm not misusing of doctor.ValidateDescriptrs?
- could you explain why the validation checker would fail on dropped descriptors?
- any ideas for how to create a backup with an invalid descriptor that would fail these checks?

File: pkg/ccl/backupccl/restore_job.go (hunk 2 above)
Comment (fqazi) [reply pkg/ccl/backupccl/restore_job.go:714]: The fact its temp table makes in interesting, do you know the exact error? Was it related to namespace entries? Wondering if its a validation bug

**Comment by msbutler** [comment]
1. sounds good!

2.
> Can you give the exact error we are seeing? Is it linked to a missing namespace entry or something else?

Once the latest TC run finishes (added a commit that doesn't filter dropped tables), you'll see a flurry of failures. As an example, running `TestBackupAndRestoreJobDescription` returns:
`pq: the restoring cluster contains invalid descriptors, %s: descriptor validation check failed: Examining 47 descriptors and 47 namespace entries...
          ParentID 104, ParentSchemaID 105: namespace entry "bank" (106): no matching name info in draining names of dropped relation`
          
Note that some fail when the restore rolls back, leading to an unclear error msg, e.g. "table [104] not found".  

3. Ideally, the test would generate the corrupt backup, so the backup is the same release as the restore, so I'll go with the pb to json manipulation route.

File: pkg/ccl/backupccl/restore_planning.go (hunk 4 above)
Comment (msbutler) [reply pkg/ccl/backupccl/restore_planning.go]: Done.

**Comment by msbutler** [comment]
hm, I'm also starting to think this PR is not really necessary. It seems sql schema code validates all descriptors anyway. If I run the unit test i created with `skip_descriptor_check`, the restore still fails with the following stack trace, which implies that the `collection.GetAllDescriptors` call will do a bunch of validation.

```
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate.(*validationErrorAccumulator).decorate
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate/validate.go:238
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate.(*validationErrorAccumulator).Report
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate/validate.go:173
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc.(*wrapper).ValidateBackReferences
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc/validate.go:251
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate.Validate.func3
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate/validate.go:101
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate.(*validationErrorAccumulator).validateDescriptorsAtLevel
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate/validate.go:189
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate.Validate
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/internal/validate/validate.go:96
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).Validate
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/validate.go:38
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).finalizeDescriptors
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:568
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.getDescriptorsByID
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:203
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).getDescriptorsByID
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/descriptor.go:104
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/sql/catalog/descs.(*Collection).GetAllDescriptors
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | 	github.com/cockroachdb/cockroach/pkg/sql/catalog/descs/collection.go:434
I221115 22:35:13.674710 362732 jobs/registry.go:1234  [nsql1] 2 +  | github.com/cockroachdb/cockroach/pkg/ccl/backupccl/backupresolver.LoadAllDescs.func1
```

File: pkg/ccl/backupccl/testdata/backup-restore/restore-check-descriptors [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+exec-sql
+SELECT crdb_internal.unsafe_delete_namespace_entry("parentID", "parentSchemaID", name, id) FROM system.namespace WHERE name = 'child'
+----
+
+job resume=a
+----
+
+job tag=a wait-for-state=revert-failed
+----
```
[the remark this answers is not in the document]
Comment (stevendanna) [reply pkg/ccl/backupccl/testdata/backup-restore/restore-check-descriptors:41]: This behaviour is intentional: https://github.com/cockroachdb/cockroach/pull/69300

The idea being that for many jobs OnFailOrCancel running to completion is required to put the system back into a usable state.

**Comment by msbutler** [comment]
I've done a bit more digging, and I don't think there's significant additional descriptor validation coverage that this PR, via `doctor.ExamineDescriptors`, provides over the automatic validation conducted whenever a descriptor is read or written from disk. 
- `doctor.ExamineDescriptors` calls `catalog.Validate` on the `validate.Write` level of validation, the highest form of validation. Every time a descriptor is written to disk, it is also validated with the `validate.Write` level. In other words, in this PR, we're currently calling this doctor cmd to validate stuff, right after we've already validated all the written descriptors! This seems useless.
- `doctor.ExamineDescriptors` also calls `ValidateDescriptorReferencesInJob`, which does not get called during automatic descriptor validation, but neither do we in this PR. The jobs table should not get checked until the very end of restore, so do we really want to fail the whole restore, after hours of work, just bc a descriptor has some bad reference to the jobs table?

File: pkg/ccl/backupccl/testdata/backup-restore/restore-check-descriptors (hunk 7 above)
[the remark this answers is not in the document]
Comment (msbutler) [reply pkg/ccl/backupccl/testdata/backup-restore/restore-check-descriptors:41]: but should this job retry _forever_? I guess with exponential backoff, the time between retry frequency will become sufficiently large, preventing the exhaustion of jobs resources.

**Comment by msbutler** [comment]
@stevendanna @fqazi does my analysis above seem sound? TLDR: i don't think we need all this extra validation. Perhaps i'll keep the test to ensure that the automatic validation does indeed catch invalid descriptor bugs

**Comment by fqazi** [comment]
@msbutler Yes it makes it sound unnecessary unless we can catch things earlier.

**Comment by msbutler** [comment]
thanks for clarifying! closing this PR and will open a seperate, test only PR, that checks that restore validation does indeed occur automatically.

**Pull request closed without merging**